File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

int main(){
    cin.tie(0);cout.tie(0);ios::sync_with_stdio(false);
    int n; ll m; cin >> n >> m;
    vector<ll> tree(n);
    ll low, high; low = high = 0;
    for(int i = 0; i < n; i++){ 
        cin >> tree[i];
        if(high < tree[i]) high = tree[i];
    }
    while(low <= high){
        ll mid = (low+high)/2;
        ll total = 0;
        for(auto t : tree){
            if(t > mid) total += t-mid;
        }
        if(total >= m) low = mid+1;
        else high = mid-1;
    }
    cout << high;
    return 0;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

def main():
    n, m = map(int, input().split())
    tree = tuple(map(int, input().split()))
    low = 0
    high = max(tree)
    while low <= high:
        mid = (low+high)//2
        total = sum(t - mid for t in tree if t > mid)
        if total >= m: low = mid+1
        else: high = mid-1
    print(high)
    return 0
if __name__ == '__main__':
    sys.exit(main())